gui/velociraptor/src/components/users/user.js

```diff
@@ -0,0 +1,271 @@
+            let permissions = Object.assign({}, prevState.permissions);
+            selected_roles[role] = !selected_roles[role];
+            if (selected_roles[role]) {
+                permissions.roles.push(role);
+            } else {
+                permissions.roles = permissions.roles.filter(p_role => p_role !== role);
+            }
+            return {selected_roles, permissions};
+        });
+    }
+
+    handlePermissionSelect = (permission) => {
+        this.setState(prevState => {
+            let permissions = Object.assign({}, prevState.permissions);
+            permissions[permission] = !permissions[permission];
+            return {permissions};
+        });
+    }
+
+    render() {
+        let permission_items = Object.assign({}, this.state.permissions);
+        delete permission_items.roles;
+
+        let nameForm;
+        if (this.state.new_user) {
+            nameForm = <FormGroup>
+                         <FormControl placeholder="Username"
+                                      aria-label="Username"
+                                      aria-describedby="basic-addon1"
+                                      onChange={(e) => this.setState({name: e.target.value})}/>
+                       </FormGroup>;
+        }
+
+        return(
+                <>
+                  <h2 className="mx-1">{this.state.new_user ? "Create a new user" : "Edit user: " + this.state.name}</h2>
+                  <div className="pt-1 col-12">
+                    <Form>
+                      { nameForm }
+                      <FormGroup>
+                        <FormControl placeholder="Password"
+                                     aria-label="Password"
+                                     type="password"
+                                     aria-describedby="basic-addon1"
+                                     onChange={(e) => this.setState({password: e.target.value})}
+                          />
+                      </FormGroup>
+                      <FormGroup>
+                        <FormControl placeholder="Email"
+                                     aria-label="Email"
+                                     aria-describedby="basic-addon1"
+                                     onChange={(e) => this.setState({email: e.target.value})}
+                                     value={this.state.email}/>
+                      </FormGroup>
+                      <FormGroup>
+                        <InputGroup className="mb-3">
+                          <DropdownList onItemToggle={(c)=>{this.handlePermissionSelect(c);}}
+                                        name="Permissions"
+                                        items={permission_items}/>
+                          <DropdownList onItemToggle={(c)=>{this.handleRoleSelect(c);}}
+                                        name="Roles"
+                                        items={this.state.selected_roles}/>
+                        </InputGroup>
```

Comment:
These appear to only allow the selection of a single item, which doesn't reflect the ACL model at all. It also doesn't display easily what the current settings are. Despite Permissions being represented by a set of booleans and Roles being represented as a list of strings, both are effectively groups of checkboxes. `VeloForm` has an `artifactset` option that could easily be extended to be a generic multi-select interface that should work well here. It would also be helpful to use friendly names for each option and provide a tooltip that describes what each option does.

Reply:
Do you mean it looks like it only allows single selection or that it doesn't allow to select multiple? Because it should allow the selection of multiple.
But okay, thanks for the hints, I'll take a look at it.

Reply:
Actually, you're correct that it does allow multiple selections - but I can't say I'm a fan of that UX